content/blog/2025-05-13-bril2wasm.md

```diff
@@ -0,0 +1,55 @@
+
+Another challenge was working with code created by another group in the class, who were working toward the same deadline we were. Throughout the course of their Brilooped project, they naturally ran into adjustments to the Brilooped format, which then cascaded into affecting our project as well; as the two were developing concurrently, we did not have a well of extensive Briloop documentation to refer back to when developing our programs. However, we did have the advantage of being able to talk to Dev and Gerardo in real time. We could, and did, ask for help directly whenever we ran into trouble, and they were really really helpful in helping us get our input from Briloop working. We’re incredibly grateful for their help throughout this process.
+
+One more challenging aspect of the project was, for the limited capacity in which we used Gemini, wrangling the LLM to produce actually correct (or even executable) code; though it still probably saved us the time it would take to do the uninteresting bit ourselves, it was certainly not an effortless “get out of jail free” card.
+
+# Analysis
+
+## Testing Methodology
+We tested our code thoroughly on small examples as we implemented core functionality. This allowed us to catch errors before our code became too complex to fix. After completing our implementation, we tested on the Bril benchmarks core test suite. To do this, we first took Briloop core benchmark translations, provided by the Briloop group, and translated them into json files using their Briloop2json command. After, we ran the files through our code to output the WebAssembly translation, then ran this translation using Wasmtime. We compared the output of our WebAssembly files to the output from the original Bril files. With this method, we were able to confirm correctness, in the sense that the WebAssembly output program behaved identically on our test cases to the original Bril programs (1 test per benchmark).
+
+Note that our WebAssembly runtime prints all arguments one per line, instead of multiple arguments on the same line. Thus, our test cases ignored whitespace differences.
+
+## Performance
+Fortunately (and perhaps unsurprisingly), the WebAssembly executed significantly faster in wasmtime than the original Bril interpreter could run Bril files. Using a script that runs all of the core benchmarks in sequence using the provided arguments, we benchmarked performance using hyperfine. The original Bril interpreter took a mean of 3.883 seconds with a standard deviation of 0.106 seconds to run through the entire folder. The WebAssembly versions took only 1.043 seconds with a standard deviation of 0.044 seconds. This is an observed speed increase of over 73.1%.
```

Comment:
Can you please include a few details of your experimental setup (hardware/OS/Deno & wasmtime versions/etc.)?